src/lang.rs

```diff
@@ -185,6 +206,164 @@ impl Expr {
+/// - None is returned
+fn expect_fn<'a, F, O, EF>(
+    mut parser: F,
+    mut error_fn: EF,
+) -> impl FnMut(Span<'a>) -> IResult<Span, Option<O>>
+where
+    F: Parser<Span<'a>, O, nom::error::Error<Span<'a>>>,
+    EF: FnMut(&'a QueryContainer<'a>, QueryRange),
+{
+    move |input: Span<'a>| match parser.parse(input) {
+        Ok((remaining, out)) => Ok((remaining, Some(out))),
+        Err(nom::Err::Error(nom::error::Error { input, .. }))
+        | Err(nom::Err::Failure(nom::error::Error { input, .. })) => {
+            let r = input.to_sync_point();
+            let end = r.end - input.location_offset();
+            error_fn(input.extra, r);
+            let next = input.slice(end..);
+            Ok((next, None))
+        }
+        Err(err) => Err(err),
+    }
+}
+
+/// A version of the `delimited()` combinator() that calls an error-handling function when the
+/// terminator parser fails.
+pub fn delimited2<'a, O1, O2, O3, F, G, H, EF>(
+    mut first: F,
+    mut second: G,
+    mut third: H,
+    mut error_fn: EF,
+) -> impl FnMut(Span<'a>) -> IResult<Span<'a>, O2, nom::error::Error<Span<'a>>>
+where
+    F: Parser<Span<'a>, O1, nom::error::Error<Span<'a>>>,
+    G: Parser<Span<'a>, O2, nom::error::Error<Span<'a>>>,
+    H: Parser<Span<'a>, O3, nom::error::Error<Span<'a>>>,
+    EF: FnMut(&'a QueryContainer<'a>, QueryRange),
+{
+    move |input: Span<'a>| {
+        let full_r = input.to_sync_point();
+        let (input, _) = first.parse(input)?;
+        let (input, o2) = second.parse(input)?;
+
+        match third.parse(input) {
+            Ok((input, _)) => Ok((input, o2)),
+            Err(_) => {
+                let start = input.location_offset();
+                let mut remaining = input.clone();
+
+                loop {
+                    if remaining.is_empty() {
+                        error_fn(
+                            remaining.extra,
+                            Range {
+                                start: full_r.start,
+                                end: remaining.location_offset(),
+                            },
+                        );
+                        return Ok((remaining, o2));
+                    }
+
+                    remaining = remaining.slice(1..);
+                    let end = remaining.location_offset();
+                    let res = third.parse(remaining);
+
+                    if let Ok((remaining, _)) = res {
+                        remaining
+                            .extra
+                            .report_error_for("unhandled input")
+                            .with_code_range(Range { start, end }, "")
+                            .send_report();
+                        return Ok((remaining, o2));
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Combinator that expects some optional whitespace followed by the vertical bar or EOF.  If that
+/// input is not found, the error message is logged.
+fn expect_bar<'a, M>(
```

Comment:
maybe rename to `expect_pipe` to avoid bar in the context of foo, bar, baz, etc.